pallets/parachain-system/src/lib.rs

```diff
@@ -296,15 +343,84 @@ pub mod pallet {
 				weight += T::DbWeight::get().writes(1);
 			}
 
+			// If `MaxUnincludedLen` is present in the storage, parachain head
+			// is always expected to be included into the relay storage proof.
+			let para_head_with_len = <MaxUnincludedLen<T>>::get().map(|max_len| {
+				let relay_chain_state = Self::relay_state_proof()
+					.expect("relay state proof must be present in storage");
+				let validation_data =
+					Self::validation_data().expect("validation data must be present in storage");
+
+				let relay_state_proof = RelayChainStateProof::new(
+					T::SelfParaId::get(),
+					validation_data.relay_parent_storage_root,
+					relay_chain_state,
+				)
+				.expect("Invalid relay chain state proof");
+				(
+					relay_state_proof
+						.read_included_para_head()
+						.expect("Invalid para head in relay chain state proof"),
+					max_len,
+				)
+			});
+			weight += T::DbWeight::get().reads(1);
+
+			// Update unincluded segment related storage values.
+			if let Some((para_head, max_len)) = para_head_with_len {
+				// Weight used for reading para head.
+				weight += T::DbWeight::get().reads(2);
+
+				if !max_len.is_zero() {
+					// Weight used by `on_finalize` in case the len is non-zero.
+					weight += T::DbWeight::get().reads_writes(4, 2);
+
+					let (dropped, left_count): (Vec<Ancestor>, u32) =
+						<UnincludedSegment<T>>::mutate(|chain| {
+							// Drop everything up to the block with an included para head, if present.
+							let idx = chain
+								.iter()
+								.position(|block| block.para_head() == &para_head)
+								.map_or(0, |idx| idx + 1); // inclusive.
+
+							let left_count = (idx..chain.len()).count() as u32;
+							let dropped = chain.drain(..idx).collect();
+							(dropped, left_count)
+						});
+					weight += T::DbWeight::get().reads_writes(1, 1);
+
+					// sanity-check there's place for the block at finalization phase.
+					//
+					// TODO: this potentially restricts parachains from decreasing `MaxUnincludedLen` value.
```

Comment:
Yes, it does - parachains manually adjusting `MaxUnincludedLen` would potentially brick their chain. This is fine as long as there is a warning in the storage item doc comment.

We will need good ways for parachains to adjust this.

Reply:
I basically referenced https://github.com/paritytech/cumulus/pull/2438#discussion_r1163461339, need to link to a github issue too before the merge